lib/version.js

```diff
@@ -46,27 +37,70 @@ function version (args, silent, cb_) {
       return cb_(er)
     }
 
-    var newVersion = semver.valid(args[0])
-    if (!newVersion) newVersion = semver.inc(data.version, args[0])
-    if (!newVersion) return cb_(version.usage)
-    if (data.version === newVersion) return cb_(new Error('Version not changed'))
-    data.version = newVersion
-    var lifecycleData = Object.create(data)
-    lifecycleData._id = data.name + '@' + newVersion
-    var localData = {}
-
-    var where = npm.prefix
-    chain([
-          [checkGit, localData],
-          [lifecycle, lifecycleData, 'preversion', where],
-          [updatePackage, newVersion, silent],
-          [lifecycle, lifecycleData, 'version', where],
-          [commit, localData, newVersion],
-          [lifecycle, lifecycleData, 'postversion', where] ],
-          cb_)
+    if (args[0] === 'from-git') {
+      retrieveTagVersion(silent, data, cb_)
+    } else {
+      var newVersion = semver.valid(args[0])
+      if (!newVersion) newVersion = semver.inc(data.version, args[0])
+      if (!newVersion) return cb_(version.usage)
+      persistVersion(newVersion, silent, data, cb_)
+    }
   })
 }
 
+function retrieveTagVersion (silent, data, cb_) {
+  chain([
+    verifyGit,
+    parseLastGitTag
+  ], function (er, results) {
+    if (er) return cb_(er)
+    var localData = {
+      hasGit: true,
+      existingTag: true
+    }
+
+    var version = results[results.length - 1]
+    persistVersion(version, silent, data, localData, cb_)
+  })
+}
+
+function parseLastGitTag (cb) {
+  var options = { env: process.env }
+  git.whichAndExec(['describe', '--abbrev=0'], options, function (er, stdout) {
+    if (er) {
+      if (er.message.indexOf('No names found') !== -1) return cb(new Error('No tags found'))
+      return cb(er)
+    }
+
+    var tag = stdout.trim()
+    var version = semver.valid(tag)
```

Comment:
I think it would probably be good to check if `tag-version-prefix` can be stripped off the tag, and do so if it can. See my other comments for why.